File: implementation/go/land.go
package main

import (
	"fmt"
)

type Land struct {
	rows int 
	cols int 
	nodes [][] *piziNode
}

func (l *Land) init(rows, cols int) {
	l.rows = rows
	l.cols = cols
	// (ابی)شب که میشه به عشق تو غزل غزل صدا میشم ترانه خون قصه تمام عاشقا میشم
	l.nodes = make([][]*piziNode, rows)
	for i := range l.nodes {
		l.nodes[i] = make([]*piziNode, cols)

		for j := range l.nodes[i] {
			l.nodes[i][j] = initNode(i, j)
		}
	}

	// this is not good (search for the bether solution)
	for i := range l.nodes {
		for j := range l.nodes[i] {
			l.detectNeighbors(l.nodes[i][j])
		}
	}
}

func (l *Land) makeObstacle(node *piziNode ,status bool) {
	node.isObstacle = status
}

// func (l *Land) makeObstacle(row, col int ,status bool) {
// 	l.nodes[row][col].isObstacle = status
// }

func initNode(i, j int) *piziNode{
	return &piziNode{
		row: i,
		col: j,
		tag: "new",
		isObstacle: false,
		next: nil,
		costToGoal: -1, // not sure about this.
	}
}

func (l *Land) detectNeighbors(node *piziNode) {
	neighbors := []*piziNode{}

	// left neighbor
	if node.row - 1 >= 0 {
		neighbors = append(neighbors, l.nodes[node.row-1][node.col])
	}
	
	// right neighbor
	if node.row + 1 < l.rows {
		neighbors = append(neighbors, l.nodes[node.row+1][node.col])
	}
	
	// upper neighbor
	if node.col - 1 >= 0 {
		neighbors = append(neighbors, l.nodes[node.row][node.col-1])
	}
	
	// lower neighbor
	if node.col + 1 < l.cols {
		neighbors = append(neighbors, l.nodes[node.row][node.col+1])
	}

	node.neighbors = neighbors
}


func (l *Land) print() {
	fmt.Println(l.nodes)
}

File: implementation/qizi/land.cpp
#include "land.h"

Land::Land(int rows, int cols)
{
    this->rows = rows;
    this->cols = cols;

    nodes = vector<vector<Node*>>(rows);
    for (int i=0; i<rows; i++) {
        nodes[i] = vector<Node*>(cols);
        for (int j=0; j<cols; j++) {
            nodes[i][j] = createNode(i, j);
        }
    }

    // bad code
    for (int i=0; i<rows; i++) {
        for (int j=0; j<cols; j++) {
            detectNeighbors(nodes[i][j]);
        }
    }
}

void Land::configureObstacle(Node* node, bool status)
{
    node->setIsObstacle(status);
}

Node *Land::createNode(int i, int j)
{
    Node* node = new Node();

    node->setRow(i);
    node->setCol(j);
    node->setIsObstacle(false);
    node->setTag("new");
    node->setNext(nullptr);
    node->setCostToGoal(-1);
    node->setKey(0);

    return node;
}

void Land::detectNeighbors(Node *node)
{
    vector<Node*> neighbors;

    // left neighbor
    if (node->getRow() - 1 >= 0) {
        neighbors.push_back(nodes[node->getRow()-1][node->getCol()]);
    }

    // right neighbor
    if (node->getRow() + 1 < rows ){
        neighbors.push_back(nodes[node->getRow()+1][node->getCol()]);
    }

    // upper neighbor
    if (node->getCol() - 1 >= 0) {
        neighbors.push_back(nodes[node->getRow()][node->getCol()-1]);
    }

    // lower neighbor
    if (node->getCol() + 1 < cols) {
        neighbors.push_back(nodes[node->getRow()][node->getCol()+1]);
    }

    node->setNeighbors(neighbors);
}

int Land::getRows() const
{
    return rows;
}

void Land::setRows(int newRows)
{
    rows = newRows;
}

int Land::getCols() const
{
    return cols;
}

void Land::setCols(int newCols)
{
    cols = newCols;
}

vector<vector<Node *> > Land::getNodes() const
{
    return nodes;
}

void Land::setNodes(const vector<vector<Node *> > &newNodes)
{
    nodes = newNodes;
}
